File: Tests/Full/Pascal/GameResources.pas
unit GameResources;

interface
	uses SysUtils, SGSDK_Core, SGSDK_Font, SGSDK_Audio, SGSDK_Graphics, SGSDK_Input, SGSDK_Physics, SGSDK_MappyLoader;

	procedure LoadResources();
	procedure FreeResources();
	function GameFont(font: String): Font;
	function GameImage(image: String): Bitmap;
	function GameSound(sound: String): SoundEffect;
	function GameMusic(music: String): Music;
	function GameMap(mapName: String): Map;
 
implementation
	var
		_Images: Array of Bitmap;
		_Fonts: Array of Font;
		_Sounds: Array of SoundEffect;
		_Music: Array of Music;
		_Maps: Array of Map;

		_ImagesStr: Array of String;
		_FontsStr: Array of String;
		_SoundsStr: Array of String;
		_MusicStr: Array of String;
		_MapsStr: Array of String;

		_Background: Bitmap;
		_Animation: Bitmap;
		_LoaderEmpty: Bitmap;
		_LoaderFull: Bitmap;
		_LoadingFont: Font;
		_StartSound: SoundEffect;

	procedure PlaySwinGameIntro();
	const
		ANI_X = 143;
		ANI_Y = 134;
		ANI_W = 546;
		ANI_H = 327;
		ANI_V_CELL_COUNT = 6;
		ANI_CELL_COUNT = 11;
	var
		i : Integer;
	begin
		PlaySoundEffect(_StartSound);
		Sleep(200);
		for i:= 0 to ANI_CELL_COUNT - 1 do
		begin
			DrawBitmap(_Background, 0, 0);
			DrawBitmapPart(_Animation, 
				(i div ANI_V_CELL_COUNT) * ANI_W, (i mod ANI_V_CELL_COUNT) * ANI_H,
				ANI_W, ANI_H, ANI_X, ANI_Y);
			RefreshScreen();
			ProcessEvents();
			Sleep(20);
		end;
		Sleep(1500);
	end;

	procedure ShowLoadingScreen();
	begin
		_Background := LoadBitmap(GetPathToResource('SplashBack.png', ImageResource));
		DrawBitmap(_Background, 0, 0);
		RefreshScreen(60);
		ProcessEvents();

		_Animation := LoadBitmap(GetPathToResource('SwinGameAni.png', ImageResource));
		_LoaderEmpty := LoadBitmap(GetPathToResource('loader_empty.png', ImageResource));
		_LoaderFull := LoadBitmap(GetPathToResource('loader_full.png', ImageResource));
		_LoadingFont := LoadFont(GetPathToResource('arial.ttf', FontResource), 12);
		_StartSound := LoadSoundEffect(GetPathToResource('SwinGameStart.ogg', SoundResource));

		PlaySwinGameIntro();
	end;

	procedure ShowMessage(message: String; number: Integer);
	const
		TX = 310; TY = 493; 
		TW = 200; TH = 25;
		STEPS = 5;
		BG_X = 279; BG_Y = 453;
	var
		fullW: Integer;
	begin
		fullW := 260 * number div STEPS;
		DrawBitmap(_LoaderEmpty, BG_X, BG_Y);
		DrawBitmapPart(_LoaderFull, 0, 0, fullW, 66, BG_X, BG_Y);

		DrawTextLines(message, ColorWhite, ColorTransparent, _LoadingFont, AlignCenter, TX, TY, TW, TH);
		RefreshScreen(60);
		ProcessEvents();
	end;

	procedure EndLoadingScreen();
	begin
		ProcessEvents();
		Sleep(500);
		ClearScreen();
		RefreshScreen(60);
		FreeFont(_LoadingFont);
		FreeBitmap(_Background);
		FreeBitmap(_Animation);
    FreeBitmap(_LoaderEmpty);
    FreeBitmap(_LoaderFull);
		FreeSoundEffect(_StartSound);
	end;
	
	procedure NewMap(mapName: String);
	begin
		SetLength(_Maps, Length(_Maps) + 1);
		SetLength(_MapsStr, Length(_MapsStr) + 1);
		_Maps[High(_Maps)] := LoadMap(mapName);
		_MapsStr[High(_MapsStr)] := mapName;
	end;
	
	procedure NewFont(fontName, fileName: String; size: Integer);
	begin
		SetLength(_Fonts, Length(_Fonts) + 1);
		SetLength(_FontsStr, Length(_FontsStr) + 1);
		_Fonts[High(_Fonts)] := LoadFont(GetPathToResource(fileName, FontResource), size);
		_FontsStr[High(_FontsStr)] := fontName;
	end;
	
	procedure NewImage(imageName, fileName: String);
	begin
		SetLength(_Images, Length(_Images) + 1);
		SetLength(_ImagesStr, Length(_ImagesStr) + 1);
		_Images[High(_Images)] := LoadBitmap(GetPathToResource(fileName, ImageResource));
		_ImagesStr[High(_ImagesStr)] := imageName;
	end;
	
	procedure NewTransparentColourImage(imageName, fileName: String; transColour: Colour);
	begin
		SetLength(_Images, Length(_Images) + 1);
		SetLength(_ImagesStr, Length(_ImagesStr) + 1);
		_Images[High(_Images)] := LoadBitmap(GetPathToResource(fileName, ImageResource), true, transColour);
		_ImagesStr[High(_ImagesStr)] := imageName;
	end;
	
	procedure NewTransparentColorImage(imageName, fileName: String; transColor: Color);
	begin
		NewTransparentColourImage(imageName, fileName, transColor);
	end;
	
	procedure NewSound(soundName, fileName: String);
	begin
		SetLength(_Sounds, Length(_Sounds) + 1);
		SetLength(_SoundsStr, Length(_SoundsStr) + 1);
		_Sounds[High(_Sounds)] := LoadSoundEffect(GetPathToResource(fileName, SoundResource));
		_SoundsStr[High(_SoundsStr)] := soundName;
	end;
	
	procedure NewMusic(musicName, fileName: String);
	begin
		SetLength(_Music, Length(_Music) + 1);
		SetLength(_MusicStr, Length(_MusicStr) + 1);
		_Music[High(_Music)] := LoadMusic(GetPathToResource(fileName, SoundResource));
		_MusicStr[High(_MusicStr)] := musicName;
	end;
	
	procedure LoadFonts();
	begin
		NewFont('ArialLarge', 'arial.ttf', 80);
		NewFont('Courier', 'cour.ttf', 15);
		NewFont('CourierLarge', 'cour.ttf', 28);
	end;

	procedure FreeFonts();
	var
		i: Integer;
	begin
		for i := Low(_Fonts) to High(_Fonts) do
		begin
			FreeFont(_Fonts[i]);
		end;
	end;

	procedure LoadImages();
	var
		i: Integer;
	begin
		NewImage('SmallBall', 'ball_small.png');
		NewImage('BallImage1', 'ball.png');
		NewImage('BallImage2', 'ball2.png');
		NewImage('SmallBall', 'ball_small.png');
		NewImage('Running', 'running.png');
		NewImage('Explosion', 'Explosion.png');
		NewImage('Ship', 'ship.png');
		NewImage('Sea', 'sea.png');
		NewImage('BGA', 'BackgroundDrawArea.png');
		NewImage('BG', 'BackgroundMain.png');
		NewImage('Frame1', 'F01.png');
		NewImage('Frame2', 'F02.png');
		NewImage('Frame3', 'F03.png');
		NewImage('Frame4', 'F04.png');
		NewImage('Frame5', 'F05.png');
		NewImage('Frame6', 'F06.png');
		NewImage('enShip', 'enShip.png');
		NewTransparentColourImage('BlueExplosion', 'explosion_pro.jpg', ColourBlack);
		for i := 0 to 39 do
		begin
			NewTransparentColourImage('Explode_' + IntToStr(i), 'explode_' + IntToStr(i) + '.jpg', ColourBlack);
		end;
		//NewImage('NoImages', 'Ufo.png');
	end;

	procedure FreeImages();
	var
		i: Integer;
	begin
		for i := Low(_Images) to High(_Images) do
		begin
			FreeBitmap(_Images[i]);
		end;
	end;

	procedure LoadSounds();
	begin
		NewSound('Shock', 'shock.wav');
		//NewSound('NoSound', 'sound.ogg');
	end;
	
	procedure LoadMaps();
	begin
		NewMap('test');
		NewMap('test3');
	end;
	
	procedure FreeSounds();
	var
		i: Integer;
	begin
		for i := Low(_Sounds) to High(_Sounds) do
		begin
			FreeSoundEffect(_Sounds[i]);
		end;
	end;

	procedure LoadMusics();
	begin
		NewMusic('Fast', 'Fast.mp3');
		//NewMusic('NoMusic', 'sound.mp3');
	end;

	procedure FreeMusics();
	var
		i: Integer;
	begin
		for i := Low(_Music) to High(_Music) do
		begin
			FreeMusic(_Music[i]);
		end;
	end;
	
	procedure FreeMaps();
	var
		i: Integer;
	begin
		for i := Low(_Maps) to High(_Maps) do
		begin
			FreeMap(_Maps[i]);
		end;
	end;

	procedure LoadResources();
	var
		oldW, oldH: Integer;
	begin
		oldW := ScreenWidth();
		oldH := ScreenHeight();

		ChangeScreenSize(800, 600);

		//Remove sleeps once "real" game resources are being loaded
		ShowLoadingScreen();
		ShowMessage('Loading fonts...', 0); 
		LoadFonts();
		Sleep(50);

		ShowMessage('Loading images...', 1);
		LoadImages();
		Sleep(50);

		ShowMessage('Loading sounds...', 2);
		LoadSounds();
		Sleep(50);

		ShowMessage('Loading music...', 3);
		LoadMusics();
		Sleep(50);
		
		ShowMessage('Loading maps...', 4);
		LoadMaps();
		Sleep(50);

		ShowMessage('Game loaded...', 5);
		Sleep(50);
		EndLoadingScreen();

		ChangeScreenSize(oldW, oldH);
	end;

	procedure FreeResources();
	begin
		FreeFonts();
		FreeImages();
		FreeMusics();
		FreeSounds();
		FreeMaps();
	end;

	function GameFont(font: String): Font; 
	var
		i: Integer;
	begin
		for i := Low(_FontsStr) to High(_FontsStr) do
		begin
			if _FontsStr[i] = font then begin
				result := _Fonts[i];
				exit;
			end;
		end;
		raise exception.create('Font ' + font + ' does not exist...');
	end;

	function GameImage(image: String): Bitmap;
	var
	i: Integer;
	begin
		for i := Low(_ImagesStr) to High(_ImagesStr) do
		begin
			if _ImagesStr[i] = image then begin
				result := _Images[i];
				exit;
			end;
		end;
		raise exception.create('Image ' + image + ' does not exist...');
	end;

	function GameSound(sound: String): SoundEffect; 
	var
		i: Integer;
	begin
		for i := Low(_SoundsStr) to High(_SoundsStr) do
		begin
			if _SoundsStr[i] = sound then begin
				result := _Sounds[i];
				exit;
			end;
		end;
		raise exception.create('Sound ' + sound + ' does not exist...');
	end;
	
	function GameMap(mapName: String): Map;
	var
		i: Integer;
	begin
		for i := Low(_MapsStr) to High(_MapsStr) do
		begin
			if _MapsStr[i] = mapName then begin
				result := _Maps[i];
				exit;
			end;
		end;
		raise exception.create('Map ' + mapName + ' does not exist...');
	end;
	
	function GameMusic(music: String): Music;
	var
		i: Integer;
	begin
		for i := Low(_MusicStr) to High(_MusicStr) do
		begin
			if _MusicStr[i] = music then begin
				result := _Music[i];
				exit;
			end;
		end;
		raise exception.create('Music ' + music + ' does not exist...');
	end;
end.

File: Documentation/BattleShips/Battleships/Battleships/GameResources.vb
Imports SwinGame
Imports System.Collections.Generic

Public Module GameResources

    Private Sub LoadFonts()
        NewFont("ArialLarge", "arial.ttf", 80)
        NewFont("Courier", "cour.ttf", 14)
        NewFont("CourierSmall", "cour.ttf", 8)
        NewFont("Menu", "ffaccess.ttf", 8)
    End Sub

    Private Sub LoadImages()
        'Backgrounds
        NewImage("Menu", "main_page.jpg")
        NewImage("Discovery", "discover.jpg")
        NewImage("Deploy", "deploy.jpg")

        'Deployment
        NewImage("LeftRightButton", "deploy_dir_button_horiz.png")
        NewImage("UpDownButton", "deploy_dir_button_vert.png")
        NewImage("SelectedShip", "deploy_button_hl.png")
        NewImage("PlayButton", "deploy_play_button.png")
        NewImage("RandomButton", "deploy_randomize_button.png")

        'Ships
        For i As Integer = 1 To 5
            NewImage("ShipLR" & i, "ship_deploy_horiz_" & i & ".png")
            NewImage("ShipUD" & i, "ship_deploy_vert_" & i & ".png")

        Next

        'Explosions
        NewImage("Explosion", "explosion.png")
        NewImage("Splash", "splash.png")

    End Sub

    Private Sub LoadSounds()
        NewSound("Error", "error.wav")
        NewSound("Hit", "hit.wav")
        NewSound("Sink", "sink.wav")
        NewSound("Miss", "sink.wav")
        NewSound("Winner", "winner.wav")
        NewSound("Lose", "lose.wav")
        NewSound("Deploy", "deploy.wav")

    End Sub

    Private Sub LoadMusic()
        NewMusic("Background", "horrordrone.mp3")
    End Sub

    Private Sub LoadMaps()
    End Sub

    ''' <summary>
    ''' Gets a Font Loaded in the Resources
    ''' </summary>
    ''' <param name="font">Name of Font</param>
    ''' <returns>The Font Loaded with this Name</returns>

    Public Function GameFont(ByVal font As String) As Font
        Return _Fonts(font)
    End Function

    ''' <summary>
    ''' Gets an Image loaded in the Resources
    ''' </summary>
    ''' <param name="image">Name of image</param>
    ''' <returns>The image loaded with this name</returns>

    Public Function GameImage(ByVal image As String) As Bitmap
        Return _Images(image)
    End Function

    ''' <summary>
    ''' Gets an sound loaded in the Resources
    ''' </summary>
    ''' <param name="sound">Name of sound</param>
    ''' <returns>The sound with this name</returns>

    Public Function GameSound(ByVal sound As String) As SoundEffect
        Return _Sounds(sound)
    End Function

    ''' <summary>
    ''' Gets the music loaded in the Resources
    ''' </summary>
    ''' <param name="music">Name of music</param>
    ''' <returns>The music with this name</returns>

    Public Function GameMusic(ByVal music As String) As Music
        Return _Music(music)
    End Function

    ''' <summary>
    ''' Gets a map loaded in the Resources
    ''' </summary>
    ''' <param name="map">Name of map</param>
    ''' <returns>The map with this name</returns>

    Public Function GameMap(ByVal map As String) As Map
        Return _Maps(map)
    End Function

    Private _Images As New Dictionary(Of String, Bitmap)
    Private _Fonts As New Dictionary(Of String, Font)
    Private _Sounds As New Dictionary(Of String, SoundEffect)
    Private _Music As New Dictionary(Of String, Music)
    Private _Maps As New Dictionary(Of String, Map)

    Private _Background As Bitmap
    Private _Animation As Bitmap
    Private _LoaderFull As Bitmap
    Private _LoaderEmpty As Bitmap
    Private _LoadingFont As Font
    Private _StartSound As SoundEffect

    ''' <summary>
    ''' The Resources Class stores all of the Games Media Resources, such as Images, Fonts
    ''' Sounds, Music, and Maps.
    ''' </summary>

    Public Sub LoadResources()
        Dim width, height As Integer

        width = Core.ScreenWidth()
        height = Core.ScreenHeight()

        Core.ChangeScreenSize(800, 600)

        ShowLoadingScreen()

        ShowMessage("Loading fonts...", 0)
        LoadFonts()
        Core.Sleep(100)

        ShowMessage("Loading images...", 1)
        LoadImages()
        Core.Sleep(100)

        ShowMessage("Loading sounds...", 2)
        LoadSounds()
        Core.Sleep(100)

        ShowMessage("Loading music...", 3)
        LoadMusic()
        Core.Sleep(100)

        ShowMessage("Loading maps...", 4)
        LoadMaps()
        Core.Sleep(100)

        Core.Sleep(100)
        ShowMessage("Game loaded...", 5)
        Core.Sleep(100)
        EndLoadingScreen(width, height)
    End Sub

    Private Sub ShowLoadingScreen()
        _Background = Graphics.LoadBitmap(Core.GetPathToResource("SplashBack.png", ResourceKind.ImageResource))
        Graphics.DrawBitmap(_Background, 0, 0)
        Core.RefreshScreen()
        Core.ProcessEvents()

        _Animation = Graphics.LoadBitmap(Core.GetPathToResource("SwinGameAni.jpg", ResourceKind.ImageResource))
        _LoadingFont = Text.LoadFont(Core.GetPathToResource("arial.ttf", ResourceKind.FontResource), 12)
        _StartSound = Audio.LoadSoundEffect(Core.GetPathToResource("SwinGameStart.ogg", ResourceKind.SoundResource))

		_LoaderFull = Graphics.LoadBitmap(Core.GetPathToResource("loader_full.png", ResourceKind.ImageResource))
		_LoaderEmpty = Graphics.LoadBitmap(Core.GetPathToResource("loader_empty.png", ResourceKind.ImageResource))

        PlaySwinGameIntro()
    End Sub

    Private Sub PlaySwinGameIntro()
        Const ANI_X As Integer = 143, ANI_Y As Integer = 134, ANI_W As Integer = 546, ANI_H As Integer = 327, ANI_V_CELL_COUNT As Integer = 6, ANI_CELL_COUNT As Integer = 11

        Audio.PlaySoundEffect(_StartSound)
        Core.Sleep(200)

        Dim i As Integer
        For i = 0 To ANI_CELL_COUNT - 1
	        Graphics.DrawBitmap(_Background, 0, 0)
	        Graphics.DrawBitmapPart(_Animation, (i \ ANI_V_CELL_COUNT) * ANI_W, (i mod ANI_V_CELL_COUNT) * ANI_H, ANI_W, ANI_H, ANI_X, ANI_Y)
	        Core.Sleep(20)
            Core.RefreshScreen()
            Core.ProcessEvents()
        Next i

        Core.Sleep(1500)

    End Sub

    Private Sub ShowMessage(ByVal message As String, ByVal number As Integer)
        Const TX As Integer = 310, TY As Integer = 493, TW As Integer = 200, TH As Integer = 25, STEPS As Integer = 5, BG_X As Integer = 279, BG_Y As Integer = 453
		
		Dim fullW As Integer

		fullW = 260 * number \ STEPS
		Graphics.DrawBitmap(_LoaderEmpty, BG_X, BG_Y)
		Graphics.DrawBitmapPart(_LoaderFull, 0, 0, fullW, 66, BG_X, BG_Y)
				
		Text.DrawTextLines(message, Color.White, Color.Transparent, _LoadingFont, FontAlignment.AlignCenter, TX, TY, TW, TH)

        Core.RefreshScreen()
        Core.ProcessEvents()
    End Sub

    Private Sub EndLoadingScreen(ByVal width As Integer, ByVal height As Integer)
        Core.ProcessEvents()
        Core.Sleep(500)
        Graphics.ClearScreen()
        Core.RefreshScreen()
        Text.FreeFont(_LoadingFont)
        Graphics.FreeBitmap(_Background)
        Graphics.FreeBitmap(_Animation)
        Graphics.FreeBitmap(_LoaderEmpty)
        Graphics.FreeBitmap(_LoaderFull)
        Audio.FreeSoundEffect(_StartSound)
        Core.ChangeScreenSize(width, height)
    End Sub

    Private Sub NewMap(ByVal mapName As String)
        _Maps.Add(mapName, MappyLoader.LoadMap(mapName))
    End Sub

    Private Sub NewFont(ByVal fontName As String, ByVal filename As String, ByVal size As Integer)
        _Fonts.Add(fontName, Text.LoadFont(Core.GetPathToResource(filename, ResourceKind.FontResource), size))
    End Sub

    Private Sub NewImage(ByVal imageName As String, ByVal filename As String)
        _Images.Add(imageName, Graphics.LoadBitmap(Core.GetPathToResource(filename, ResourceKind.ImageResource)))
    End Sub

    Private Sub NewTransparentColorImage(ByVal imageName As String, ByVal fileName As String, ByVal transColor As Color)
        _Images.Add(imageName, Graphics.LoadBitmap(Core.GetPathToResource(fileName, ResourceKind.ImageResource), True, transColor))
    End Sub

    Private Sub NewTransparentColourImage(ByVal imageName As String, ByVal fileName As String, ByVal transColor As Color)
        NewTransparentColorImage(imageName, fileName, transColor)
    End Sub

    Private Sub NewSound(ByVal soundName As String, ByVal filename As String)
        _Sounds.Add(soundName, Audio.LoadSoundEffect(Core.GetPathToResource(filename, ResourceKind.SoundResource)))
    End Sub

    Private Sub NewMusic(ByVal musicName As String, ByVal filename As String)
        _Music.Add(musicName, Audio.LoadMusic(Core.GetPathToResource(filename, ResourceKind.SoundResource)))
    End Sub

    Private Sub FreeFonts()
        Dim obj As Font
        For Each obj In _Fonts.Values
            Text.FreeFont(obj)
        Next
    End Sub

    Private Sub FreeImages()
        Dim obj As Bitmap
        For Each obj In _Images.Values
            Graphics.FreeBitmap(obj)
        Next
    End Sub

    Private Sub FreeSounds()
        Dim obj As SoundEffect
        For Each obj In _Sounds.Values
            Audio.FreeSoundEffect(obj)
        Next
    End Sub

    Private Sub FreeMusic()
        Dim obj As Music
        For Each obj In _Music.Values
            Audio.FreeMusic(obj)
        Next
    End Sub

    Private Sub FreeMaps()
        Dim obj As Map
        For Each obj In _Maps.Values
            MappyLoader.FreeMap(obj)
        Next
    End Sub

    Public Sub FreeResources()
        FreeFonts()
        FreeImages()
        FreeMusic()
        FreeSounds()
        FreeMaps()
		Core.ProcessEvents()
    End Sub
End Module
